src/Servers/Kestrel/test/FunctionalTests/UnixDomainSocketsTests.cs

```diff
@@ -115,6 +117,67 @@ async Task EchoServer(ConnectionContext connection)
             }
         }
 
+#if LIBUV
+        [OSSkipCondition(OperatingSystems.Windows, SkipReason = "Libuv does not support unix domain sockets on Windows.")]
+#else
+        [OSSkipCondition(OperatingSystems.Windows, WindowsVersions.Win7, WindowsVersions.Win8, WindowsVersions.Win81, WindowsVersions.Win2008R2, SkipReason = "UnixDomainSocketEndPoint is not supported on older versions of Windows")]
+#endif
+        [ConditionalFact]
+        [CollectDump]
+        public async Task TestUnixDomainSocketWithUrl()
+        {
+            var path = Path.GetTempFileName();
+            var url = $"http://unix:/{path}";
+
+            Delete(path);
+
+            try
+            {
+                var hostBuilder = TransportSelector.GetWebHostBuilder()
+                    .UseUrls(url)
+                    .UseKestrel()
+                    .ConfigureServices(AddTestLogging)
+                    .Configure(app =>
+                    {
+                        app.Run(async context =>
+                        {
+                            await context.Response.WriteAsync("Hello World");
+                        });
+                    });
+
+                using (var host = hostBuilder.Build())
+                {
+                    await host.StartAsync();
+
+                    // https://github.com/dotnet/corefx/issues/5999
+                    // .NET Core HttpClient does not support unix sockets, it's difficult to parse raw response data. below is a little hacky way.
+                    using (var socket = new Socket(AddressFamily.Unix, SocketType.Stream, ProtocolType.Unspecified))
+                    {
+                        await socket.ConnectAsync(new UnixDomainSocketEndPoint(path));
+
+                        byte[] readBuffer = new byte[32];
+                        var httpRequest = Encoding.ASCII.GetBytes("GET / HTTP/1.1\r\nHost:\r\n\r\n");
+                        await socket.SendAsync(httpRequest, SocketFlags.None);
+
+                        await socket.ReceiveAsync(readBuffer, SocketFlags.None);
```

Comment:
@mjohn Sorry for the delay. This change looks really good, and we'd like to merge it soon. I have a few changes I'd like to make to this test to make it less flaky and less likely to hang.

The above line could return an incomplete read buffer leading to flaky tests. Can call ReceiveAsync in a loop until the expected amount of data is read or 0 is returned like the [above test](https://github.com/aspnet/AspNetCore/pull/12112/files#diff-a69a4830c53b497fe9f2eb25a73247ddL96-L103)?

The above test should check if ReceiveAsync returned 0 and exit the loop but doesn't. If you want to fix that at the same time great. If not, we can clean that up ourselves later. Every time we're awaiting a Task or ValueTask in a test, we should also be using the `.DefaultTimeout()` extension methods.

Reply:
@halter73 Thanks for your comment, I tried to make your suggestion but ReceiveAsync didn't return 0. It hangs forever and receives timeout error. I'll push the branch and if I get over this error I can also fix the above test you mentioned.

I'm not sure whether the http respone size will vary. I also read http 1.1 specifications to determine response but chunked transfer encoding isn't easy as i thought.

Reply:
In this case, we wouldn't expect ReceiveAsync to actually ever return 0. Since you're making an `HTTP/1.1` request without a `Connection: close` header, Kestrel assumes it's a keepalive request. The easiest fix is probably just add the  `Connection: close` to the request.